### ISSUE TITLE
Validation fails for milestone releases

### ISSUE THREAD

**Issue by markt-asf** [eclipse-ee4j/glassfish-spec-version-maven-plugin#14]
Noticed with Jakarta WebSocket but expected to impact all projects using the plugin when attempting to produce a milestone build or any other build with a release qualifier.

The reported error is:
```
[INFO] --- spec-version-maven-plugin:2.0:check-module (default) @ jakarta.websocket-client-api ---

[ jakarta.websocket:jakarta.websocket-client-api:2.0.0-M1 ]
{ groupIdPrefix=jakarta. spec-version=2.0 apiPackage=jakarta.websocket.client API final spec-impl-version=2.0.0-M1 }
- WARNING: Bundle-Version (2.0.0.M1) should be 2.0.0-M1
```

The issue is that OSGi version numbers always use `.` as the separator, even for qualifiers so a version number of `2.0.0-M1` becomes `2.0.0.M1` for OSGi but the plug-in does not account for this.

Currently working around the issue with `-DignoreErrors`

**Comment by romain-grecourt** [comment]
Can you provide the full configuration that leads to this error ?
E.g. I see final spec, but that seems odd for a milestone build.

**Comment by markt-asf** [comment]
Everything you need should be here:
https://ci.eclipse.org/websocket/job/1_websocket-api-build-and-stage/35/

We've only just started doing milestone builds so I'm not surprised some of the version numbering is slightly inconsistent. We're really only worried about getting milestone API JARs out there so dependent projects can start their migration from javax.* to jakarta.*

**Comment by romain-grecourt** [comment]
The plugin complete configuration of the plugin isn't in the build log.

**Comment by markt-asf** [comment]
https://github.com/eclipse-ee4j/websocket-api/blob/master/api/pom.xml#L211
https://github.com/eclipse-ee4j/websocket-api/blob/master/api/server/pom.xml#L47
https://github.com/eclipse-ee4j/websocket-api/blob/master/api/client/pom.xml#L47

Note that the CI job switches the version from 2.0.0-SNAPSHOT to 2.0.0-M1

**Comment by smillidge** [comment]
Got the same problem trying to do a 2.0.0-RC1 build of Jakarta Concurrency as part of Jakarta EE 9. I'm happy to fix the plugin if somebody can give me a hint. 

Error message while trying to release an RC build to maven central is;

```
[INFO] --- spec-version-maven-plugin:2.0:set-spec-properties (default) @ jakarta.enterprise.concurrent-api ---
[INFO] 
[INFO] -- spec properties --
[INFO] spec.implementation.version = 2.0.0-RC1
[INFO] spec.specification.version = 2.0
[INFO] spec.bundle.symbolic-name = jakarta.enterprise.concurrent-api
[INFO] spec.bundle.spec.version = 2.0
[INFO] spec.extension.name = jakarta.enterprise.concurrent
[INFO] spec.bundle.version = 2.0.0-RC1
...
...

...
...
[INFO] --- spec-version-maven-plugin:2.0:check-module (default) @ jakarta.enterprise.concurrent-api ---

[ jakarta.enterprise.concurrent:jakarta.enterprise.concurrent-api:2.0.0-RC1 ]
{ groupIdPrefix=jakarta. spec-version=2.0 apiPackage=jakarta.enterprise.concurrent API final spec-impl-version=2.0.0-RC1 }
- WARNING: Bundle-Version (2.0.0.RC1) should be 2.0.0-RC1
```

**Comment by bshannon** [comment]
@romain-grecourt can you help with this?  I'm going to assign it to you for now,
but feel free to assign it to Steve or someone else if you don't have the time.

**Comment by arjantijms** [comment]
@bshannon @romain-grecourt and others:

I'll take a look at this (mentioned to Romain already). 

As the `-RCx` qualifier is the one we settled on, and since I think it should only be a Maven thing, not show up in the bundle, I'll specifically strip this one then. In effect it will behave just like `-SNAPSHOT` does today. 

I'm hesitant to do this for all qualifiers, as currently there's code looking at e.g. the `-bx` qualifier and doing special stuff for `nonFinal` builds (like adding `.99.` in between).

**Comment by bshannon** [comment]
I think the goal was to turn all the release qualifiers into numbers for OSGi.
At the time we were only thinking about -bXX qualifiers.  Ideally we would
define a sort order between all the -Bxx, -RCxx, and -Mxx qualifiers, but
that seems like too much work.  Ignoring the sort order. we could map
(e.g.) 2.3.0-M1 to the OSGi number 2.2.98.1, and map 2.3.0-RC1 to 2.2.97.1,
or something like that.  The point is to come up with an OSGi number that is
clearly "less than" the original unqualified number (2.3.0 in the example, assumed
to always end in ".0") by backing up to the previous release number and using
a new micro release number that's large enough to be unlikely to have been
used by a real release.

**Comment by arjantijms** [comment]
> we could map (e.g.) 2.3.0-M1 to the OSGi number 2.2.98.1, and map 2.3.0-RC1 to 2.2.97.1

In practice that might not work, as many OSGi constraints are set on the minor number, so e.g. "2.2>". Meaning it would work with the faked release candidate for the new number but not with the actual release.

Such a mapping can still be useful for early builds I guess (M builds?), but I'm less sure about RC ones, where the idea is that they can be promoted to final without much or any other changes.

**Comment by romain-grecourt** [comment]
Perhaps this should only be allowed when `nonFinal` is `false` ?

**Comment by arjantijms** [comment]
The "simple" RCx case is fixed via #24. I guess we should leave this open to discuss the support for additional milestone semantics?

**Comment by arjantijms** [comment]
> Perhaps this should only be allowed when nonFinal is false ?

That might be a good option. I asked a question about whether to retain the `nonFinal` mode in #23.

**Comment by romain-grecourt** [comment]
I commented on that issue, the versioning [document](https://wiki.eclipse.org/JakartaEE_Maven_Versioning_Rules) still mentions about final and non final.

**Comment by bshannon** [comment]
The mapping to OSGi is never going to be "right".  You're never going to know
whether 2.3.0-M2 is supposed to be less than or greater than 2.3.0-RC1.
OSGi version comparisons for non-final artifacts just have to compare as less
then the corresponding final artifact; anything else doesn't really matter.
(Although if you only ever use -Bxx they should all compare as expected.)

**Comment by KyleAure** [comment]
Ran into this trying to do the concurrency M1 release. 
```
[INFO] --- spec-version:2.1:check-module (default) @ jakarta.enterprise.concurrent-api ---

[ jakarta.enterprise.concurrent:jakarta.enterprise.concurrent-api:3.1.0-M1 ]
{ groupIdPrefix=jakarta. spec-version=3.1 apiPackage=jakarta.enterprise.concurrent API final spec-impl-version=3.1.0-M1 }
- WARNING: Bundle-Version (3.1.0.M1) should be 3.1.0-M1
```

**Comment by arjantijms** [comment]
> Ran into this trying to do the concurrency M1 release.
> 
> ```
> [INFO] --- spec-version:2.1:check-module (default) @ jakarta.enterprise.concurrent-api ---
> 
> [ jakarta.enterprise.concurrent:jakarta.enterprise.concurrent-api:3.1.0-M1 ]
> { groupIdPrefix=jakarta. spec-version=3.1 apiPackage=jakarta.enterprise.concurrent API final spec-impl-version=3.1.0-M1 }
> - WARNING: Bundle-Version (3.1.0.M1) should be 3.1.0-M1
> ```
> 
> Seeing as how all Jakarta EE specs are required to do an M1 release by the end of November this is likely going to break a lot release builds 🙅🏼

True, in all our wisdom we decided at the time that the qualifier to be used was "RCx" as commented above and implemented  in #24. See also https://github.com/eclipse-ee4j/glassfish-spec-version-maven-plugin/blob/master/src/main/java/org/glassfish/spec/Artifact.java#L61

Nobody ever followed up about supporting "Mx" qualifiers.
